File: is52038b-labs/13/java/HeapIndexFunctionsTest.java
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import java.util.Random;
import org.junit.Test;
import org.junit.runner.JUnitCore;
import org.junit.runner.Result;

public class HeapIndexFunctionsTest {
    @Test
    public void testLeft() {
        assertEquals("what is the index of the left child of the node at position 0?", 1, Heap.left(0));
        assertEquals("what is the index of the left child of the node at position 1?", 3, Heap.left(1));
        assertEquals("what is the index of the left child of the node at position 2?", 5, Heap.left(2));
        Random r = new Random();
        for (int i = 0; i < 100; i++) {
            int x = r.nextInt(32768);
            assertEquals("a random choice of parent has an erroneous child!", 2*x+1, Heap.left(x));
        }
    }
    @Test
    public void testRight() {
        assertEquals("what is the index of the right child of the node at position 0?", 2, Heap.right(0));
        assertEquals("what is the index of the right child of the node at position 1?", 4, Heap.right(1));
        assertEquals("what is the index of the right child of the node at position 2?", 6, Heap.right(2));
        Random r = new Random();
        for (int i = 0; i < 100; i++) {
            int x = r.nextInt(32768);
            assertEquals("a random choice of parent has an erroneous child!", 2*x+2, Heap.right(x));
        }
    }
    @Test
    public void testParent() {
        assertEquals("what is the index of the parent of the node at position 1?", 0, Heap.parent(1));
        assertEquals("what is the index of the parent of the node at position 2?", 0, Heap.parent(2));
        assertEquals("what is the index of the parent of the node at position 3?", 1, Heap.parent(3));
        Random r = new Random();
        for (int i = 0; i < 100; i++) {
            int x = 1 + r.nextInt(32767);
            assertEquals("a random choice of child has an erroneous parent!", (x-1)/2, Heap.parent(x));
        }
    }
}

File: completed-labs/13/cpp/HeapIndexFunctionsTest.cpp
#include <cmath>
#include <cppunit/extensions/HelperMacros.h>

#include "Heap.hpp"

class HeapIndexFunctionsTest : public CppUnit::TestFixture {
  CPPUNIT_TEST_SUITE(HeapIndexFunctionsTest);
  CPPUNIT_TEST(testLeft);
  CPPUNIT_TEST(testRight);
  CPPUNIT_TEST(testParent);
  CPPUNIT_TEST_SUITE_END();

public:
  void setUp() {};
  void tearDown() {};

  void testLeft();
  void testRight();
  void testParent();
};

void HeapIndexFunctionsTest::testLeft() {
  CPPUNIT_ASSERT_EQUAL_MESSAGE("what is the index of the left child of the node at position 0?", 1U, left(0));
  CPPUNIT_ASSERT_EQUAL_MESSAGE("what is the index of the left child of the node at position 1?", 3U, left(1));
  CPPUNIT_ASSERT_EQUAL_MESSAGE("what is the index of the left child of the node at position 2?", 5U, left(2));
  for (int i = 0; i < 100; i++) {
    unsigned int x = rand() % 32768;
    CPPUNIT_ASSERT_EQUAL_MESSAGE("a random choice of parent has an erroneous child!", 2*x+1, left(x));
  }
}

void HeapIndexFunctionsTest::testRight() {
  CPPUNIT_ASSERT_EQUAL_MESSAGE("what is the index of the right child of the node at position 0?", 2U, right(0));
  CPPUNIT_ASSERT_EQUAL_MESSAGE("what is the index of the right child of the node at position 1?", 4U, right(1));
  CPPUNIT_ASSERT_EQUAL_MESSAGE("what is the index of the right child of the node at position 2?", 6U, right(2));
  for (int i = 0; i < 100; i++) {
    unsigned int x = rand() % 32768;
    CPPUNIT_ASSERT_EQUAL_MESSAGE("a random choice of parent has an erroneous child!", 2*x+2, right(x));
  }
}

void HeapIndexFunctionsTest::testParent() {
  CPPUNIT_ASSERT_EQUAL_MESSAGE("what is the index of the parent of the node at position 1?", 0U, parent(1));
  CPPUNIT_ASSERT_EQUAL_MESSAGE("what is the index of the parent of the node at position 2?", 0U, parent(2));
  CPPUNIT_ASSERT_EQUAL_MESSAGE("what is the index of the parent of the node at position 3?", 1U, parent(3));
  for (int i = 0; i < 100; i++) {
    unsigned int x = 1 + rand() % 32767;
    CPPUNIT_ASSERT_EQUAL_MESSAGE("a random choice of child has an erroneous parent!", (x-1)/2, parent(x));
  }
}

CPPUNIT_TEST_SUITE_NAMED_REGISTRATION(HeapIndexFunctionsTest, "IndexFunctions");
CPPUNIT_REGISTRY_ADD_TO_DEFAULT("IndexFunctions");
